### PR TITLE
Setup DistNeighborloader for graph store sampling

### PR DESCRIPTION
**Pull request by kmontemayor2-sc**
**Scope of work done**

Setting up distributed neighborloader for graph store (GS) mode.

In this change we:

1. Introduce `RemoteDistDataset` as an input for the loader
2. Allow the loader to take in `list[Tensor]` as inputs for GS sampling.
3. Add tests that the above work together :)

Where is the documentation for this feature?: N/A

Did you add automated tests or write a test plan?

***Updated Changelog.md?*** NO

***Ready for code review?:*** NO

### CODE REVIEW

File: python/gigl/distributed/distributed_neighborloader.py [hunk 1]
```diff
@@ -71,12 +82,15 @@ def __init__(
...
-                indices of seed nodes to start sampling from.
+            input_nodes (Tenor | Tuple[NodeType, Tenor] | list[Tenor] | Tuple[NodeType, list[Tenor]]):
+                The nodes to start sampling from.
                 It is of type `torch.LongTensor` for homogeneous graphs.
                 If set to `None` for homogeneous settings, all nodes will be considered.
                 In heterogeneous graphs, this flag must be passed in as a tuple that holds
                 the node type and node indices. (default: `None`)
+                For Graph Store mode, this must be a tuple of (NodeType, list[Tenor]) or list[Tenor].
+                Where each Tensor in the list is the node ids to sample from, for each server.
+                e.g. [[10, 20], [30, 40]] means sample from nodes 10 and 20 on server 0, and nodes 30 and 40 on server 1.
```
Comment (svij-sc) [inline python/gigl/distributed/distributed_neighborloader.py]: Since this is not a ragged tensor, I guess we have to sample the same amount from every server?
i.e. if there are a 100 servers the clients will distribute all servers data evenly among all the clients and sample accordingly?

Wouldnt this induce a high fanout networking?

File: python/gigl/distributed/distributed_neighborloader.py [hunk 2]
```diff
@@ -3,7 +3,11 @@
 
 import torch
 from graphlearn_torch.channel import SampleMessage
-from graphlearn_torch.distributed import DistLoader, MpDistSamplingWorkerOptions
+from graphlearn_torch.distributed import (
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: I'll make similar changes - just instead of `list[Tensor]` it'll be `list[tuple[Tensor, Tensor, Optional[Tensor]]]` where the tuple is `anchors, pos labels, neg labels` and we add some new util to `RemoteDistDataset` to fetch them from server :)

[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: (Since this PR is already on the large end - I wasn't planning on having those changes in here yet.)

File: python/gigl/distributed/distributed_neighborloader.py [hunk 3]
```diff
@@ -37,10 +43,15 @@
...
+        dataset: Union[DistDataset, RemoteDistDataset],
         num_neighbors: Union[list[int], dict[EdgeType, list[int]]],
         input_nodes: Optional[
-            Union[torch.Tensor, Tuple[NodeType, torch.Tensor]]
+            Union[
+                torch.Tensor,
+                Tuple[NodeType, torch.Tensor],
+                list[torch.Tensor],
+                Tuple[NodeType, list[torch.Tensor]],
+            ]
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: Myself and Matt aligned offline after many conversations with potential users.

The two main routes here are:

1. Have users pass in the nodes to sample from explicetely
2. Inside the DistLoader fetch the nodes to sample (e.g. query `RemoteDistDataset` directly)

Our decision here was to allow more freedom for users to pass in the data they want, it's possible that the "anchors" (or labels) may not be part of the dataset strictly, and so it should be possible for users to be explicit with what data they want to use.

Also I do think this was in the doc I had sent out.

File: python/gigl/distributed/distributed_neighborloader.py [hunk 4]
```diff
@@ -62,7 +73,7 @@ def __init__(
         https://pytorch-geometric.readthedocs.io/en/2.5.2/_modules/torch_geometric/distributed/dist_neighbor_loader.html#DistNeighborLoader
 
         Args:
-            dataset (DistDataset): The dataset to sample from.
+            dataset (DistDataset | RemoteDistDataset): The dataset to sample from. Must be a "RemoteDistDataset" if using Graph Store mode.
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: Ack, done.

File: python/gigl/distributed/distributed_neighborloader.py [hunk 5]
```diff
@@ -71,12 +82,15 @@ def __init__(
             context (deprecated - will be removed soon) (DistributedContext): Distributed context information of the current process.
             local_process_rank (deprecated - will be removed soon) (int): Required if context provided. The local rank of the current process within a node.
             local_process_world_size (deprecated - will be removed soon)(int): Required if context provided. The total number of processes within a node.
-            input_nodes (torch.Tensor or Tuple[str, torch.Tensor]): The
-                indices of seed nodes to start sampling from.
+            input_nodes (Tenor | Tuple[NodeType, Tenor] | list[Tenor] | Tuple[NodeType, list[Tenor]]):
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: Fixed, thanks.

File: python/gigl/distributed/distributed_neighborloader.py [hunk 6]
```diff
@@ -188,10 +202,219 @@ def __init__(
                 local_process_rank=local_rank
             )
         )
+
+        # Determines if the node ids passed in are heterogeneous or homogeneous.
+        self._is_labeled_heterogeneous = False
+        if isinstance(dataset, DistDataset):
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: Good idea! let me do that :)

File: python/gigl/distributed/distributed_neighborloader.py [hunk 7]
```diff
@@ -188,10 +202,219 @@ def __init__(
...
+                dataset,
+                num_workers,
+            )
+
+        self._is_labeled_heterogeneous = dataset_metadata.is_labeled_heterogeneous
+        self._node_feature_info = dataset_metadata.node_feature_info
+        self._edge_feature_info = dataset_metadata.edge_feature_info
+
+        num_neighbors = patch_fanout_for_sampling(
+            list(dataset_metadata.edge_feature_info.keys())
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: Sure, updated.

File: python/gigl/distributed/utils/neighborloader.py [hunk 8]
```diff
@@ -15,6 +16,22 @@
 _GraphType = TypeVar("_GraphType", Data, HeteroData)
 
 
+@dataclass(frozen=True)
+class DatasetMetadata:
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/utils/neighborloader.py]: Good idea, done!

File: python/gigl/distributed/distributed_neighborloader.py [hunk 9]
```diff
@@ -188,10 +202,219 @@ def __init__(
...
+            # Note that each compute node may have multiple connections to each storage node, once per compute process.
+            # E.g. if there are 4 gpus per compute node, then there will be 4 connections to each storage node.
+            # We need to this because if we don't, then there is a race condition when initalizing the samplers on the storage nodes [1]
+            # Where since the lock is per *server* (e.g. per storage node), if we try to start one connection from compute node 0, and compute node 1
+            # Then we deadlock and fail.
+            # [1]: https://github.com/alibaba/graphlearn-for-pytorch/blob/main/graphlearn_torch/python/distributed/dist_server.py#L129-L167
+            node_rank = dataset.cluster_info.compute_node_rank
+            for target_node_rank in range(dataset.cluster_info.num_compute_nodes):
+                if node_rank == target_node_rank:
+                    super().__init__(
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: Updated the big comment above with where the call fails, let me know if that explains anything :)

[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: Per dataloader code instantiation (e.g. user writes `DistNeighborLoader(...)`), of which there will be COMPUTE_WORLD_SIZE (num compute nodes * num processes per compute) total instantiations.

Then, each compute node (e.g. GPU machine), there should be one "set" of connections (e.g. if we have 4 gpus/compute a set will be of size 4) to each storage node.

I've added ascii art above here, let me know if this is more clear.

File: python/gigl/distributed/utils/neighborloader.py [hunk 10]
```diff
@@ -15,6 +16,22 @@
...
+    """
+    Shared metadata between the local and remote datasets.
+    """
+
+    # If the dataset is labeled heterogeneous. E.g. one node type, one edge type, and "label" edges.
+    is_labeled_heterogeneous: bool
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/utils/neighborloader.py]: We use "labeled_heterogeneous" elsewhere already. [1]

If we want to update the name I think that's fine (thoguh I prefer the shorter name myself :P) but let's do that in a separate PR?

[1]: https://github.com/Snapchat/GiGL/blob/main/python/gigl/distributed/distributed_neighborloader.py#L340

File: python/gigl/distributed/distributed_neighborloader.py [hunk 11]
```diff
@@ -188,10 +202,219 @@ def __init__(
...
+                raise ValueError(
+                    "When using Graph Store mode, edge feature info must be provided for heterogeneous graphs."
+                )
+            elif (
+                len(edge_feature_info) == 1
+                and DEFAULT_HOMOGENEOUS_EDGE_TYPE in edge_feature_info
+            ):
+                input_type: NodeType | None = DEFAULT_HOMOGENEOUS_NODE_TYPE
+            else:
+                input_type = fallback_input_type
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: Good callout, done.

File: python/gigl/distributed/utils/networking.py [hunk 12]
```diff
@@ -155,7 +168,8 @@ def get_internal_ip_from_node(
         # Other nodes will receive the master's IP via broadcast
         ip_list = [None]
 
-    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    device = torch.device("cpu" if not torch.cuda.is_available() else "cuda")
```
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/distributed/utils/networking.py]: Good catch, reverted.

File: python/gigl/distributed/graph_store/remote_dist_dataset.py
Comment (kmontemayor2-sc) [inline python/gigl/distributed/graph_store/remote_dist_dataset.py]: Changes being reviewed in https://github.com/Snapchat/GiGL/pull/435 :)

File: python/gigl/distributed/distributed_neighborloader.py [hunk 13]
```diff
@@ -180,18 +198,283 @@ def __init__(
...
+            nodes = input_nodes[1]
+            fallback_input_type = input_nodes[0]
+            require_edge_feature_info = True
+        else:
+            raise ValueError(
+                f"When using Graph Store mode, input nodes must be of type (list[torch.Tensor] | (NodeType, list[torch.Tensor])), received {type(input_nodes)}"
+            )
+
+        # Determine input_type based on edge_feature_info
+        if isinstance(edge_feature_info, dict):
```
Comment (mkolodner-sc) [inline python/gigl/distributed/distributed_neighborloader.py]: To double check, we should still be using edge feature info here and not dataset.get_edge_types?

**Review by mkolodner-sc — APPROVED** [review]
Thanks Kyle! Generally LGTM to me

File: python/gigl/distributed/distributed_neighborloader.py (hunk 1 above)
Comment (svij-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: From a distributed systems perspective this will create a huge bottleneck; with every compute node talking to every other compute node (train) and every storage node (data fetching).

If we are not open to addressing this preemptively, lets add a task on the backlog to evaluate this given the biggest tenet we have is cost - and if this causes bottlenecks it will increase cost.

Btw, if we do end up doing ^ optimization it might be good to deal with ragged tensor concern now.
Could we just do the following:
```
Dict[int, Tensor] | Tuple[NodeType, Dict[int, Tensor]]
```

This way we can do something as simple as the following if you just want to fetch from one storage node:
`input_nodes = (USER_NODE_TYPE, {0 : [10, 20, 30, 40]}) ` 


vs. right now you would have to do something like:
` input_nodes = (USER_NODE_TYPE, [10, 20, 30, 40], [-1, -1, -1, -1], [-1, -1, -1, -1, ], ...)`

File: python/gigl/distributed/distributed_neighborloader.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (svij-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: thank you, this helped clarify

**Review by svij-sc — COMMENTED** [review]
Thanks for addressing everything

LGTM beyond this thread:
https://github.com/Snapchat/GiGL/pull/432/changes#r2687675025

File: python/gigl/distributed/distributed_neighborloader.py (hunk 1 above)
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: > with every compute node talking to every other compute node (train) 

This is not the case - We have some setup where every compute node talks to every storage node for sampling.

The compute nodes may talk to each other for e.g. DDP, but they do not talk to each other for sampling.


> This way we can do something as simple as the following if you just want to fetch from one storage node:
input_nodes = (USER_NODE_TYPE, {0 : [10, 20, 30, 40]}) 

I think we should be able to just do `[[10, 20, 30, 40], [], [], []]` but I can verify later.

File: python/gigl/distributed/distributed_neighborloader.py (hunk 13 above)
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: updated, thanks!

File: python/gigl/distributed/distributed_neighborloader.py (hunk 1 above)
Comment (kmontemayor2-sc) [reply python/gigl/distributed/distributed_neighborloader.py]: Created https://github.com/Snapchat/GiGL/issues/440 to look into this :)

**Review by svij-sc — APPROVED** [review]

**Comment by svij-sc** [comment]
I will approve to unblock, but lets revisit the dictionary consideration here before calling this complete.
